### DISCUSSION TITLE
Need to filter properties from object diffs

### DISCUSSION THREAD

**Discussion by sailingKieler** [vitest-dev/vitest#7092] [Ideas]
Hi friends of vitest,

I'm working a lot on and with [Langium](https://github.com/eclipse-langium/langium) and Vitest has been huge benefit there.

I just ran into a problem that I would like to compare two objects for equality like
```
  expect(A).toBe(B);
```
`A` and `B` be a part of a rather large spider web of objects, und in case of a mismatch the creation of the diff visualization makes the test execution crash after some time of polluting the memory.

I saw that you have a config/customization hook for sorting the properties of objects, but no opportunity for filtering them. Ignoring a few known properties being prone to causing problems would help me a lot.
https://github.com/vitest-dev/vitest/blob/4e60333dc7235704f96314c34ca510e3901fe61f/packages/utils/src/diff/index.ts#L186-L190

Ideally I would love if I could contribute a custom `plugin` to the list of `pretty-format` plugins:
https://github.com/vitest-dev/vitest/blob/4e60333dc7235704f96314c34ca510e3901fe61f/packages/utils/src/diff/index.ts#L37-L57

Realizing that `getFormatOptions(...)` simply ignores my custom plugin contribution almost made me biting my table 😖
https://github.com/vitest-dev/vitest/blob/4e60333dc7235704f96314c34ca510e3901fe61f/packages/utils/src/diff/index.ts#L180-L191

What do think about opening something in that corner?

**Comment by hi-ogawa** [comment]
> `A` and `B` be a part of a rather large spider web of objects, und in case of a mismatch the creation of the diff visualization makes the test execution crash after some time of polluting the memory.

Do you have reproduction? I've also seen oom crash when equality fails https://github.com/vitest-dev/vitest/issues/7009
I was guessing this is a bug of printer and cyclic references, or even if it's genuinely too big, it shouldn't crash and need a way to bail out printing early somehow.

Then if this turns out to be not fixable, perhaps we can consider adding user-land workaround.

**Reply by sailingKieler** [reply]
We've public examples of [Langium](https://github.com/eclipse-langium/langium)-based projects (see also our playground https://langium.org/showcase/), I'll see whether I can provide you something. I also saw your cyclic ref detection and wondered why it didn't cut it (literally), but I didn't go deeper there.

**Reply by hi-ogawa** [reply]
FYI, I'm making a PR to help some cases https://github.com/vitest-dev/vitest/pull/7133

Not sure about your case, but it might also help avoiding a large object diff by implementing `toJSON` as it's prioritized in some cases.

**Comment by bgoscinski** [comment]
I think I have similar usecase. I'm testing a Vue application and I had a test that kept OOMing. I reduced the repro to [this](https://stackblitz.com/edit/vitest-dev-vitest-44rnhmcu?file=test%2Fsuite.test.js&initialPath=__vitest__/). The test prints HUGE diff because it traverses the whole reactivity graph where something simpler (such as `RefImpl { value: 0 }`) would completely suffice. When I increased the `CHAIN_LENGTH` constant to 100 I gave up waiting for the test result after 4 minutes when the memory consumption reached about 4GB.

I don't expect vitest to come out of the box with pretty printer that somehow knows how to print vue Refs but I'd like to teach it how to do it - some API similar to `expect.addSnapshotSerializer` would be perfect 🤔

**Reply by hi-ogawa** [reply]
Currently users can only customize via `toJSON` method, but I think we can explore other API to customize diff too. Please feel free to raise an issue / feature request.

At the same time, OOM or crash in any case should be avoided, so fixing such worst case is also important though (e.g. automatically truncate by detecting during large diff printing).

---

Maybe, this is the way to patch `toJSON` currently https://stackblitz.com/edit/vitest-dev-vitest-rh6fqqwt?file=test%2Fsuite.test.js

```js
  computed(() => 0).constructor.prototype.toJSON = function () {
    return { type: this.constructor.name, value: this.value };
  };
```

**Reply by hi-ogawa** [reply]
Also there might be several issues with a circular object. Either `toEqual/toBe` assertion itself is buggy (or too expensive) or diff generation on error is too large. Normally both happens for the same case, but the former is a clear bug and should be fixed https://github.com/vitest-dev/vitest/issues/7429.

**Reply by bgoscinski** [reply]
toJSON seems to work if I compare it to another object but when I expect something of a different type such as number it prints the whole object again: [here](https://stackblitz.com/edit/vitest-dev-vitest-ggnvak9c?file=test%2Fsuite.test.js) I'll raise an issue

**Reply by bgoscinski** [reply]
done #7478
